crates/quantization/src/rabitq/mod.rs

```diff
@@ -0,0 +1,187 @@
+use std::ops::Div;
+
+use self::operator::OperatorRaBitQ;
+use crate::reranker::error_based::ErrorBasedReranker;
+use base::index::{RaBitQuantizationOptions, SearchOptions, VectorOptions};
+use base::operator::{Borrowed, Owned, Scalar};
+use base::scalar::{ScalarLike, F32};
+use base::search::{Reranker, Vectors};
+use base::vector::{VectorBorrowed, VectorOwned};
+
+use num_traits::{Float, One, Zero};
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+
+pub mod operator;
+
+const EPSILON: f32 = 1.9;
+const THETA_LOG_DIM: u32 = 4;
+const DEFAULT_X_DOT_PRODUCT: f32 = 0.8;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct RaBitQuantizer<O: OperatorRaBitQ> {
+    dim: u32,
+    dim_pad_64: u32,
+    projection: Vec<Vec<Scalar<O>>>,
+    binary_vec_x: Vec<Vec<u64>>,
+    distance_to_centroid_square: Vec<Scalar<O>>,
+    dot_product_x: Vec<Scalar<O>>,
+    rand_bias: Vec<Scalar<O>>,
+    error_bound: Vec<Scalar<O>>,
+    factor_ip: Vec<Scalar<O>>,
+    factor_ppc: Vec<Scalar<O>>,
```

Comment:
These members should be persistent in a memory-mapped file. They are identical to `codes` for scalar & product quantization.